docs/lazy-adr/adr-006-row-propagation.md

```diff
@@ -0,0 +1,56 @@
+# ADR 006: Block Propagation with Rows
+
+## Changelog
+* 24.06.2021
+
+## Context
+It's a long story of relations between Celestia, Tendermint, and consensus block gossiping. Celestia's team discussed multiple ideas, several ADRs were made, and nothing yet was finalized. The ADR is another attempt to bring changes into block gossiping and hopefully successful.
+
+Currently, we inherit the following from Tendermint. Our codebase relies on the blocks Parts notion, which is a piece of an entire serialized block. Those Parts are gossiped between nodes in consensus and committed with PartSetHeader containing a Merkle Root of the Parts. However, Parts gossiping wasn't designed for Celestia blocks.
+
+Celestia comes with a different block representation from Tendermint. It lays out block as a table of data shares, where Rows or Columns can be and should be gossiped instead of Parts, keeping only one system-wide commitment to data.
+
+## Alternative Approaches
+### ["nah it works just don't touch it"](https://https://ahseeit.com//king-include/uploads/2020/11/121269295_375504380484919_2997236194077828589_n-6586327691.jpg) approach
+
+It turns out that we could fully treat the Tendermint consensus as a black box. However, that means that we need to verify block integrity twice for both PartSetHeader and DAHeader.
+
+#### Pros
+* Less work
+
+#### Cons
+* This isn't adequate in terms of protocol and software design
+* Potentially introduces security issues(implementer can forget/miss secondary data verification)
+* Wastes more CPU cycles on building and verifying additional Merkle Tree
+* Extends DOSing vector for huge blocks cause to verify a block, you need two verify its two complete RAM copies in different layouts.
+* We would also need to add PartSetHeader to Celestia-specs and bring there such an ugliness
+
+## Decision
+The decision is to treat Tendermint's consensus still and its gossipping as a black box, but with few amendments:
+* Entirely removing PartSetHeader, as redundant data commitment
+* Placing DAHeader instead of PartSetHeader
+* Implementing RowSet and mimicking it to PartSet
+* Removing PartSet
+
+## Detailed Design
+Fortunately, the above decision does not affect any public API, and changes are solely internal, mostly in `consensus` package, but others are also slightly affected.
+
+The essential part is to implement RowSet. Mimicking it to RowSet decreases the number of changes, as we can swap it with PartSet. However, the initial implementation will send fully erasured data rows to peers, leaving a gap for improvement by sending actual data only later. The current state of the rsmt2d library does not have any API to work with rows, only with the entire data square. Without row/col level erasure-coding API, RowSet would not check the integrity of a non-erasured Row against DAHeader, but that wouldn't be problematic to improve later.
```

Comment:
> However, the initial implementation will send fully erasured data rows to peers

Does this mean initially, you plan to send all rows (namely the full square)? So we sacrifice some bandwidth (roughly 4x the current data)? IMO that is OK for an initial implementation but we should already roughly understand and communicate the impact of sending only the ODS rows (not the full EDS) if that is the long-term plan.

Reply:
There's no need to send the entire row, just the original data. The extended data be re-computed locally.

Reply:
@musalbas weren't you previously advocating of gossiping the erasure code data too? I currently can't find the comment but this comes close: https://github.com/tendermint/spec/issues/78#issuecomment-584052848
The tendermint team is also investigating erasure coding during consensus (purely for block propagation): https://github.com/tendermint/spec/issues/277#issuecomment-809210777 but it isn't clear what direct benefits this would bring over the current approach (at least for common block sizes in app-specific chains).

But as mentioned above, I agree that we shouldn't just send the full (erasured) rows. It would be OK as intermediate step in the implementation only but we should think through the full idea before doing that first implementation step. Just sending the full rows without thinking it through doesn't seem like a good idea.

@Wondertan do we have a good intuition how switching from small chunks to larger (rows) changes the gossiping properties? Previously there is a lot of small messages that quickly disperse the network. We don't need to dive deeper if we don't change that as we have pretty good empirical evidence that this works (some multi-billion dollar networks working like a charm with this approach). I feel if we switch to much larger chunks aka rows we should do at least some napkin math to estimate the implications here which we can later validate with experiments (this is related to #438).

Reply:
@liamsi, good point.

So currently, Parts are propagated, and their size maxes to 64KiB. Also, the max message size is 1MiB. Let's now take a Row of the largest 128x128 block in comparison. The actual data size in such a case for the Row would be 128x256(shares_per_row*share_size)=32KiB. If taking the approach of sending redundant data per row, that would be 64Kib per row, as the X-axis grows and twice more Rows two sent. So my intuition here is that we wouldn't lose anything if we start sending redundant data first, as that would be exactly the same amount of data as per Part. And once we move forward with rsmt2d limitations(which wouldn't be a huge problem to fix), we even get smaller sizes of gossiped chunks. 

Are there any points I might be missing here?

Reply:
> There's no need to send the entire row, just the original data. The extended data be re-computed locally.

@musalbas, totally, but it's currently impossible due to rsmt2d API limitations, as described in the doc. We surely can update rsmt2d first and then implement sending actual data only, or we can send redundant data first and then update the implementation to extended rsmt2d API, as proposed. I like the second approach because it would be easy to implement such an upgrade while not blocking the work on rsmt2d API design and discussion. The only thing that needs to be done after rsmt2d is updated - is to update the RowSet implementation accordingly, everything else wouldn't be touched at all.

Reply:
You don't really need to update the rsmt2d implementation for this. Decoding a single row is straightforward and doesn't really need rsmt2d, it's just standard Reed-Solomon code decoding. You just need to call Decode on the row data: https://github.com/celestiaorg/rsmt2d/blob/master/infectiousRSGF8.go#L53
And probably verify that the Merkle root matches.

Implementing this as a formal API endpoint in rsmt2d, if it's even necessary, would be very trivial. Sending redundant coded data isn't an acceptable alternative IMO.

Reply:
Mostly agree with Mustafa here. But I do think if this is just an intermediate step and if we have _a very clear trajectory how to move forward_ (laid out in the ADR ideally), it would be OK. What is not acceptable is that the ADR does not clearly state how this trajectory looks like exactly. As mentioned it should be really straightforward to do this anyways. Using RS without rsmt2d could work as a bandaid / intermediate step too. We should avoid that because if we change the Codec when using rsmt2d, the place we manually encode would not be affected by that (I see future us trying to figure what is causing a weird bug otherwise).

Just a minor amendment: to locally compute the extended row data, I think calling Encode on the original row data instead of Decode as stated above (and see if merkle roots match) would be easier to understand IMO: https://github.com/celestiaorg/rsmt2d/blob/62f5a18e6b134793afca1ce284739ebbc5f2dcd7/infectiousRSGF8.go#L23
It's what creates the extension of the row in this step: https://github.com/celestiaorg/rsmt2d/blob/62f5a18e6b134793afca1ce284739ebbc5f2dcd7/extendeddatasquare.go#L75-L77

Coming back to Parts vs Rows: cool, so for a 128x128 block rows are exactly the same size (or half the size) as parts and we don't really change any properties. But the interesting observation is that row sizes vary with the block size. Maybe that is better than the fixed size chunks. I don't know yet. It certainly should be thought through a bit more as well (pros and cons should be part of this ADR).

In case of fixed sized chunks (that can be validated incrementally/individually), the receiving end could drop the connection immediately after receiving a first (small and fixed size) incorrect chunk. And the thing is that we do not really max out at 4 MB blocks (which would be the above 128x128 blocks you mention. above) but the expectation is that block will be larger (think of blocks that would rather be a few 100 MB or even much larger). Though the cool thing is that we include the DAHeader in the Proposal and whoever receives it, can simply derive the row-size from it. So validators can discard any other messages with a non-matching row size. (We need to make sure that this still works with peers that are still on an old height when we [send them data for catchup](https://github.com/celestiaorg/lazyledger-core/blob/f33b14629330e4fde47a557ed36cf70a7da1c0ce/consensus/reactor.go#L522-L539).)

BTW, this constant is never used:
https://github.com/celestiaorg/lazyledger-core/blob/280cc82d7e9e2d698c0ee8623279cab6f102ee19/types/consts/consts.go#L27-L32
We should instead make this a parameter that can be changed (either by governance or by some state logic adapting to  observed past block sizes somehow). @adlerjohn is this something we should add to the spec?